### DISCUSSION TITLE
Problem with extremely large collision hull

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#192] [Q&A] [answered]
We have a few extremely large objects in the game (like planets and moons, the maximum coordinate is 2e8) where building a ConvexHullShape fails with this error:
`Hull building failed, point 811 had an error of 1.11818e+08 (relative to tolerance: 1.11817e+11)`
Is this a case where I should increase ConvexHullShapeSettings::mHullTolerance? Or should I shrink the hull to a more reasonable size and wrap it in a ScaledShape? Or does this actually indicate some kind of problem with the input points?

**Comment by jrouwe** [comment]
I never tested Jolt with these extreme values. Are these hulls (roughly) centered around 0? You could indeed try to create a smaller hull and then scale it up, but with objects of this size, you may run into other issues in the simulation later (there are a number of 'epsilons' in the code and not all of them are exposed).

If you give me the points that you're giving to the hull builder I can take a look if I can see what the issue in the builder is.

**Reply by DocAce** [reply]
It's almost exactly centered in the two largest axes, but not along the third axis, which is several orders of magnutides smaller. I can give you the points tomorrow (after verifying that they also break the same way in the Jolt app)

BTW, this body belongs to a celestial body and won't be used for any dynamics. It's only used for raycasts (sun lens flares) and collision checks against spheres (check if the player is getting close).

**Reply by DocAce** [reply]
Yep, it gives the same error in the Jolt project. I renamed the file from cpp to .txt so I could attach it here.
[ConvexHullShapeTest.txt](https://github.com/jrouwe/JoltPhysics/files/9007937/ConvexHullShapeTest.txt)

**Reply by jrouwe** [reply]
Thanks! Unfortunately I don't have time tonight to look at it, so it will be Thursday or Friday.

**Reply by DocAce** [reply]
That's no problem, I appreciate you taking the time 👍

**Reply by DocAce** [reply]
I actually found another input mesh that causes the same problem, and it's for a much smaller object (a few hundred units). It's significantly more points, though - let me know if it's useful to you and I'll send them over, too.

**Comment by jrouwe** [comment, marked as answer]
Interesting hull you have, it's a nearly flat disc with 4337 points with an interesting pattern in it:

![image](https://user-images.githubusercontent.com/1621693/176759717-17642b96-4e27-4803-b093-8a9b098bc41b.png)

The main thing that fails is that the points are just far enough apart in the Y direction that the 'is this a 2D hull?' check thinks that it is a 3D hull. Then it tries to build a 3D hull but, because all points are nearly coplanar, and because there are so many of them, the hull building fails due to floating point precision problems.

Scaling the hull down doesn't help because it's still nearly but not quite co-planar. If I set the Y coordinate of all points to the same value, the hull building succeeds because then it goes through the 2D path.

I'm wondering if this hull is the intended hull (to me it looks like this is supposed to be a planet but maybe accidentally squashed to 2d)?

**Reply by DocAce** [reply]
It's actually a planet's rings :)
Could I somehow make the code use the 2D path if the Y extents are below a certain threshold relative to X and Z?

**Reply by jrouwe** [reply]
There's a magic constant in code:

https://github.com/jrouwe/JoltPhysics/blob/46d123f7ff3efaa8976e10162f74388fd45599b3/Jolt/Geometry/ConvexHullBuilder.cpp#L402

that makes this decision. It would be possible to expose it (and then you can make it larger), but it is risky to change it as Jolt does not touch the provided positions and then you'll end up with a hull that's built in 2D but the points are not actually 2D which can cause issues in collision detection.

**Reply by DocAce** [reply]
Thanks. For this specific case I'm probably going to solve it differently (ask for this geometry to be flagged in a way that makes the code create a CylinderShape instead of a ConvexHullShape).

The other shape with this error certainly isn't flat, though - full extents are ~ 668x367x1141, the center is relatively close to the origin. Any chance you could find out what's going wrong with this one?
[ConvexHullShapeTest.txt](https://github.com/jrouwe/JoltPhysics/files/9026955/ConvexHullShapeTest.txt)

**Reply by jrouwe** [reply]
I checked this hull and it works fine if you increase the hull tolerance from 1.0e-3 to 1.0e-2 (I don't think you need millimeter precision on an object this large). Floats have a relative accuracy of about 1e-7 so with an object of size 1141 that means the best accuracy you can get is in the order of 1e-4, this is simply too close to the default tolerance of 10^-3. 

This is the hull with a tolerance of 1e-3:

![image](https://user-images.githubusercontent.com/1621693/176886616-3d236767-64b8-4404-b94c-f6ef8e0fe841.png)

The blue arrows indicate where it went wrong. There are a number of nearly colinear points on the surface of the hull that created a sliver.

With a tolerance of 1e-2 it works:

![image](https://user-images.githubusercontent.com/1621693/176886727-4fa69091-7d10-4571-9a9c-d64dcdad0b2a.png)

If you have objects with very different sizes you could use something like max(1e-3, 1e-5 * largest extent) as the value for tolerance.

**Reply by jrouwe** [reply]
B.t.w. can I keep these hulls and make them part of a unit test? (If I have some more time I'll see if I can get the sliver detection code better)

**Reply by DocAce** [reply]
Thanks a ton for the help :)
And I cleared with my boss that it's fine for you to use the hulls.
